src/oracles/ThenaOracle.sol

```diff
@@ -0,0 +1,178 @@
+// SPDX-License-Identifier: AGPL-3.0
+pragma solidity ^0.8.13;
+
+import {Owned} from "solmate/auth/Owned.sol";
+import {FixedPointMathLib} from "solmate/utils/FixedPointMathLib.sol";
+
+import {IOracle} from "../interfaces/IOracle.sol";
+import {IThenaPair} from "../interfaces/IThenaPair.sol";
+
+/// @title Oracle using Thena TWAP oracle as data source
+/// @author zefram.eth/lookee
+/// @notice The oracle contract that provides the current price to purchase
+/// the underlying token while exercising options. Uses Thena TWAP oracle
+/// as data source, and then applies a multiplier & lower bound.
+/// Furthermore, the payment token and the underlying token must use 18 decimals.
+/// This is because the Thena oracle returns the TWAP value in 18 decimals
+/// and the OptionsToken contract also expects 18 decimals.
+contract ThenaOracle is IOracle, Owned {
+    /// -----------------------------------------------------------------------
+    /// Library usage
+    /// -----------------------------------------------------------------------
+
+    using FixedPointMathLib for uint256;
+
+    /// -----------------------------------------------------------------------
+    /// Errors
+    /// -----------------------------------------------------------------------
+
+    error ThenaOracle__StablePairsUnsupported();
+    error ThenaOracle__Overflow();
+    error ThenaOracle__BelowMinPrice();
+
+    /// -----------------------------------------------------------------------
+    /// Events
+    /// -----------------------------------------------------------------------
+
+    event SetParams(bool isToken0, uint16 multiplier, uint56 secs, uint128 minPrice);
+
+    /// -----------------------------------------------------------------------
+    /// Constants
+    /// -----------------------------------------------------------------------
+
+    /// @notice The denominator for converting the multiplier into a decimal number.
+    /// i.e. multiplier uses 4 decimals.
+    uint256 internal constant MULTIPLIER_DENOM = 10000;
+
+    /// -----------------------------------------------------------------------
+    /// Immutable parameters
+    /// -----------------------------------------------------------------------
+
+    /// @notice The Thena TWAP oracle contract (usually a pool with oracle support)
+    IThenaPair public immutable thenaPair;
+
+    /// -----------------------------------------------------------------------
+    /// Storage variables
+    /// -----------------------------------------------------------------------
+
+    /// @notice The multiplier applied to the TWAP value. Encodes the discount of
+    /// the options token. Uses 4 decimals.
+    uint16 public multiplier;
```

Comment:
consider making multiplier context specific for exercising contract rather than enshrining it within the oracle